### PR TITLE
Try add Post Type Archive in Link UI and Navigation Link via new Post Type Archive Search Handler

### LINKED ISSUE #31452: Add Post Type Archive variation to `core/navigation-link` block
## What problem does this address?

Currently you cannot add a Post Type Archive link to the Navigation block. However you can do this via the currently Menus screen. 

[Here's how you add a Post Type Archive to a menu.](https://www.wpbeginner.com/wp-tutorials/how-to-add-post-type-archive-in-wordpress-navigation-menus/)

## What is your proposed solution?

Add a `post-type-archive` block variation for the `core/navigation-link` block. We'll also need to add test coverage for this in the navigation editor.

### PR DESCRIPTION
**Pull request by getdave**
## What?

Alternative to https://github.com/WordPress/gutenberg/pull/67793

Enables searching for Post Type Archives using the Link UI and is specifically aimed at the Navigation block to bring parity with the Classic Menus system.

Closes https://github.com/WordPress/gutenberg/issues/31452

## Why?

Parity with Classic Menus where you could search for and add links to the Post Type Archives.

This also ensures that when visiting the Archives pages the `current-menu-item` class is added to the menu item.

## How?

- Adds REST Search Handler for Post Type Archives.
- Add necessary wiring for this to be used when making search requests in the Editor
- Wire up correct Navigation Link block variation.

## Testing Instructions

- Register several _Custom_ Post Types (I used the code below)

<details>
<summary>Code to register CPTs for testing</summary>

```php
function register_custom_post_types() {
	$post_types = array(
		'books'    => array(
			'singular'  => 'Book',
			'plural'    => 'Books',
			'menu_icon' => 'dashicons-book-alt',
		),
		'people'   => array(
			'singular'  => 'Person',
			'plural'    => 'People',
			'menu_icon' => 'dashicons-admin-users',
		),
		'movies'   => array(
			'singular'  => 'Movie',
			'plural'    => 'Movies',
			'menu_icon' => 'dashicons-format-video',
		),
		'products' => array(
			'singular'  => 'Product',
			'plural'    => 'Products',
			'menu_icon' => 'dashicons-cart',
		),
		'events'   => array(
			'singular'  => 'Event',
			'plural'    => 'Events',
			'menu_icon' => 'dashicons-calendar',
		),
	);

	foreach ( $post_types as $post_type => $details ) {
		$labels = array(
			'name'               => _x( $details['plural'], 'post type general name', 'textdomain' ),
			'singular_name'      => _x( $details['singular'], 'post type singular name', 'textdomain' ),
			'menu_name'          => _x( $details['plural'], 'admin menu', 'textdomain' ),
			'name_admin_bar'     => _x( $details['singular'], 'add new on admin bar', 'textdomain' ),
			'add_new'            => _x( 'Add New', strtolower( $details['singular'] ), 'textdomain' ),
			'add_new_item'       => __( 'Add New ' . $details['singular'], 'textdomain' ),
			'new_item'           => __( 'New ' . $details['singular'], 'textdomain' ),
			'edit_item'          => __( 'Edit ' . $details['singular'], 'textdomain' ),
			'view_item'          => __( 'View ' . $details['singular'], 'textdomain' ),
			'all_items'          => __( 'All ' . $details['plural'], 'textdomain' ),
			'search_items'       => __( 'Search ' . $details['plural'], 'textdomain' ),
			'parent_item_colon'  => __( 'Parent ' . $details['plural'] . ':', 'textdomain' ),
			'not_found'          => __( 'No ' . strtolower( $details['plural'] ) . ' found.', 'textdomain' ),
			'not_found_in_trash' => __( 'No ' . strtolower( $details['plural'] ) . ' found in Trash.', 'textdomain' ),
		);

		$args = array(
			'labels'        => $labels,
			'public'        => true,
			'has_archive'   => true,
			'rewrite'       => array( 'slug' => $post_type ),
			'supports'      => array( 'title', 'editor', 'thumbnail', 'excerpt', 'comments' ),
			'menu_position' => 5,
			'menu_icon'     => $details['menu_icon'],
			'show_in_rest'  => true,
		);

		register_post_type( $post_type, $args );
	}
}
add_action( 'init', 'register_custom_post_types' );
```
</details>

- Add a Navigation block
- Insert a new link item and search for one of the CPTs you registered
- See results archives in search suggestions
- Add a new link item and click `+ Add block` and then look for the `Post Type Archive Link` block variation. (hint if you can't see it then press `Browse All` and search. It will come up automatically for subsequent rounds of testing).
- Once the block variation is added, in the resulting Link UI popup see 3 suggestions for Archives. 
- Try search for another archive that isn't shown.

## Screenshots or screencast 

Screen Shot 2024-11-07 at 13 09 35

Screen Shot 2024-11-07 at 13 09 19

Screen Shot 2024-11-08 at 19 26 40

### CODE REVIEW

File: lib/experimental/class-wp-rest-post-archive-search-handler.php [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	 *
+	 * @param WP_REST_Request $request Full REST request.
+	 * @return array {
+	 *     Associative array containing found IDs and total count for the matching search results.
+	 *
+	 *     @type int[]               $ids   Found post archive IDs.
+	 *     @type string|int|WP_Error $total Numeric string containing the number of post-type archives found, or WP_Error object.
+	 * }
+	 */
+	public function search_items( WP_REST_Request $request ) {
```
Comment (getdave) [inline lib/experimental/class-wp-rest-post-archive-search-handler.php]: Need to check possible `$request` params for the search handler endpoints and ensure we're handling those correctly.

For example, offset and pagination...etc.

File: lib/experimental/class-wp-rest-post-archive-search-handler.php [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+	/**
+	 * Prepares links for the search result of a given ID.
+	 *
+	 * @since 5.6.0
+	 *
+	 * @param int $id Item ID.
+	 * @return array[] Array of link arrays for the given item.
+	 */
+	public function prepare_item_links( $id ) {
```
Comment (getdave) [inline lib/experimental/class-wp-rest-post-archive-search-handler.php]: Is it worth linking to the collection for each post archive returned? I have no use case for this so might be unnecessary at this stage,

File: lib/load.php [hunk 3]
```diff
@@ -49,6 +49,7 @@ function gutenberg_is_experiment_enabled( $name ) {
 	// WordPress 6.8 compat.
 	require __DIR__ . '/compat/wordpress-6.8/block-comments.php';
 	require __DIR__ . '/compat/wordpress-6.8/class-gutenberg-rest-comment-controller-6-8.php';
+	require_once __DIR__ . '/experimental/class-wp-rest-post-archive-search-handler.php';
```
Comment (getdave) [inline lib/load.php]: Need to put this in the correct directory.

**Comment by getdave** [comment]
Obviously we'll need to write some unit tests for the new endpoint.

**Comment by fabiankaegy** [comment]
Thank you for working on this ❤️ Really looking forward to having this in core 🚀

File: lib/experimental/class-wp-rest-post-archive-search-handler.php (hunk 1 above)
Comment (getdave) [reply lib/experimental/class-wp-rest-post-archive-search-handler.php]: Here are the [Search endpoint docs](https://developer.wordpress.org/rest-api/reference/search-results/).

We're handling pagination based on [this example from post-formats](https://github.com/WordPress/wordpress-develop/blob/8f42bad4cdcb5fb6c4ff96fcf050fb29ff6d30bf/src/wp-includes/rest-api/search/class-wp-rest-post-format-search-handler.php#L84).

**Comment by getdave** [comment]
@cr0ybot As the contributor who provided the most popular work-around for the original Issue I wonder if you would consider reviewing this PR? You can [use Playground to make this super simple](https://playground.wordpress.net/?gutenberg-pr=66821#%7B%22landingPage%22:%22/wp-admin/%22,%22features%22:%7B%22networking%22:true%7D,%22steps%22:%5B%7B%22step%22:%22login%22,%22username%22:%22admin%22,%22password%22:%22password%22%7D%5D%7D).

**Comment by TimothyBJacobs** [comment]
This doesn't really semantically feel like a search handler. We aren't querying the database at all. 

It seems like this should be using the types endpoint. It's probably worth looking at including non show_in_rest post types at this point.

**Comment by getdave** [comment]
> This doesn't really semantically feel like a search handler. We aren't querying the database at all.
> 
> It seems like this should be using the types endpoint. It's probably worth looking at including non show_in_rest post types at this point.

👋 Thanks for taking a look at this. I did look at a few options before implementing a custom search handler. [The `v2/types` endpoint](https://developer.wordpress.org/rest-api/reference/post-types/) has a few shortcomings for our requirements:

- it doesn't include a link to the archive page (I checked `_links`). I checked the `has_archive` and despite documentation to the contrary it always seems to be a boolean and not a string.
- it is not queryable by string - for example a request to `wp/v2/types/movies` will return results but `wp/v2/types/movie` will 404. Similarly requests to `wp/v2/types` lists _all_ types which still leaves us needing to filter the result client-side for the correct substring

The only way I can see us being able to us the Types endpoint is by:

- augmenting the endpoint to include the link to the archive 
- using JavaScript to filter by substring rather than doing this PHP-side (that breaks the model used in all other endpoints)

Also worth noting that we already have [the Post Formats handler](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api/search/class-wp-rest-post-format-search-handler.php) which is conceptually pretty similar to this Post Type Archive handler.

I'm open to further advice on the REST API here @TimothyBJacobs. What's the best way for a client-side application to: 

- retrieve the _href_ for the Post Type Archives for _all_ post types (with `has_archive: true`)?
- filter the Post Type Archives by a given search string?

**Comment by getdave** [comment]
> and the second worked well also until I changed the permalink structure, which broke the link in the navigation block.

@cr0ybot That's a [known issue](https://github.com/WordPress/gutenberg/issues/18345) but one I believe is separate to this PR.

File: lib/compat/wordpress-6.8/class-gutenberg-rest-post-archive-search-handler.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	 * Searches post-type archives for a given search request.
+	 *
+	 * @since 6.8.0
+	 *
+	 * @param WP_REST_Request $request Full REST request.
+	 * @return array {
+	 *     Associative array containing found IDs and total count for the matching search results.
+	 *
+	 *     @type int[]               $ids   Found post archive IDs.
+	 *     @type string|int|WP_Error $total Numeric string containing the number of post-type archives found, or WP_Error object.
```
Comment (anton-vlasenko) [inline lib/compat/wordpress-6.8/class-gutenberg-rest-post-archive-search-handler.php]:
```suggestion
	 *     @type int $total The number of post-type archives found.
```

To my knowledge, `count()` [returns](https://www.php.net/manual/en/function.count.php) an integer.

File: packages/block-library/src/navigation-link/index.php [hunk 5]
```diff
@@ -392,6 +393,27 @@ function block_core_navigation_link_build_variations() {
 				$variations[] = $variation;
 			}
 		}
+
+		// If any of the post types have `has_archive` set to true then add a post-type-archive variation.
+		$has_archive = array_filter(
+			$post_types,
+			function ( $post_type ) {
```
Comment (anton-vlasenko) [inline packages/block-library/src/navigation-link/index.php:400]: This closure can be declared as static (might improve performance).
```suggestion
			static function ( $post_type ) {
```

File: lib/compat/wordpress-6.8/class-gutenberg-rest-post-archive-search-handler.php [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	 * Prepares the search result for a given post archive ID.
+	 *
+	 * @since 6.8.0
+	 *
+	 * @param int   $id     Term ID.
+	 * @param array $fields Fields to include for the post archive.
+	 * @return array {
+	 *     Associative array containing fields for the post-archive based on the `$fields` parameter.
+	 *
+	 *     @type string    $id    Optional. Post Archive Slug.
```
Comment (anton-vlasenko) [inline lib/compat/wordpress-6.8/class-gutenberg-rest-post-archive-search-handler.php]: Nitpick.
```suggestion
	 *     @type id     $id    Optional. Post Archive Slug.
```

File: lib/experimental/rest-api.php [hunk 7]
```diff
@@ -20,6 +20,9 @@ function gutenberg_register_block_editor_settings() {
 add_action( 'rest_api_init', 'gutenberg_register_block_editor_settings' );
 
 
+
```
Comment (anton-vlasenko) [inline lib/experimental/rest-api.php:23]: Nitpick: Are these empty lines needed for some reason?

**Review by anton-vlasenko — COMMENTED** [review]
Please find my review comments below. Thank you!  
It would also be helpful to have some PHPUnit tests for the proposed new REST API controller, if possible. This isn’t a blocker, but unit tests can be useful in identifying any remaining issues, if any.

**Comment by getdave** [comment]
Thank you @anton-vlasenko 🙇 

> unit tests can be useful in identifying any remaining issues, if any

I agree 💯  I'm going to wait on writing them until I get a confidence check from @TimothyBJacobs that my reasoning for using a new handler is justified.

**Comment by getdave** [comment]
@TimothyBJacobs Do you think you might have a chance to r[eview my response to the concerns you raised](https://github.com/WordPress/gutenberg/pull/66821#issuecomment-2469663125)? Not a top priority, I'd like to get this Issue unblocked and resolved in 6.8 if possible. Very much appreciated.

**Comment by getdave** [comment]
Here's the approach I'm suggesting taking here. I will raise an alternative PR which attempts to utilise the `Types` API to achieve something similar. 

We can then review the options and get one or other of them merged so that folks can enjoy adding Post Type Archive links wherever they need to via the Link UI.

**Comment by getdave** [comment]
@TimothyBJacobs I've raised https://github.com/WordPress/gutenberg/pull/67793 which attempts to utilise the existing Types endpoint. I'd be grateful if you would consider reviewing the two PRs and letting me know which seems more appropriate based on your experience with the REST API.

**Comment by getdave** [comment]
Given the approval on https://github.com/WordPress/gutenberg/pull/67793#pullrequestreview-2511208049 I'm going to close this one out in favour of https://github.com/WordPress/gutenberg/pull/67793.

**Pull request closed without merging**